server/api_conversation.go

```diff
@@ -0,0 +1,158 @@
+		c.AbortWithError(http.StatusForbidden, errors.New("forbidden"))
+		return
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get bot: %w", err))
+		return
+	}
+
+	var reqData ConversationRequest
+	if err := json.NewDecoder(c.Request.Body).Decode(&reqData); err != nil {
+		c.AbortWithError(http.StatusBadRequest, err)
+		return
+	}
+	defer c.Request.Body.Close()
+
+	// Validation
+	if reqData.BotName == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty bot"))
+		return
+	}
+
+	bot := p.GetBotByUsername(reqData.BotName)
+	if bot == nil {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid bot name"))
+		return
+	}
+
+	post := reqData.Request
+
+	if post == nil {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid request"))
+		return
+	}
+
+	if post.Message == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty message"))
+		return
+	}
+
+	if post.ChannelId == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty channel id"))
+		return
+	}
+
+	channel, err := p.pluginAPI.Channel.Get(post.ChannelId)
+	if errors.Is(err, pluginapi.ErrNotFound) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("channel not found"))
+		return
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get channel: %w", err))
+		return
+	}
+
+	if post.UserId == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty user id"))
+		return
+	}
+
+	postingUser, err := p.pluginAPI.User.Get(post.UserId)
+	if errors.Is(err, pluginapi.ErrNotFound) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("user not found"))
+		return
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get posting user: %w", err))
+		return
+	}
+
+	// Don't respond to ourselves
+	if p.IsAnyBot(post.UserId) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("not responding to ourselves"))
+		return
+	}
+
+	list := &model.PostList{
+		Order: make([]string, 0, len(reqData.Thread)+1),
+		Posts: make(map[string]*model.Post, len(reqData.Thread)+1),
+	}
+	list.Order = append(list.Order, post.Id)
+	list.Posts[post.Id] = post
+	for i, post := range reqData.Thread {
+		list.Order = append(list.Order, post.Id)
+		list.Posts[post.Id] = reqData.Thread[i]
+	}
+
+	threadData, err := p.getMetadataForPosts(list)
+	if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get thread data: %w", err))
+		return
+	}
+
+	prompt, err := p.prompts.ChatCompletion(ai.PromptDirectMessageQuestion, p.MakeConversationContext(bot, postingUser, channel, post))
+	if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to generate prompt: %w", err))
+		return
+	}
+	prompt.AppendConversation(p.ThreadToBotConversation(bot, threadData.Posts))
+
+	// Overriding post role if requested.
+	if reqData.UseSystemRole {
```

Comment:
Sorry I think I was unclear about what this means. The system prompt is special instructions to the LLM that are weighted more highly then user instructions. Usually there is only one system prompt. You can look at OpenAI's explanation of what it means here: https://platform.openai.com/docs/guides/text-generation/chat-completions-api

I guess what I really want is to be able to call this API from other plugins and be able to customize what prompt is provided on line 122. So that we can provide some custom instructions to the LLM about how the rest of the conversation should be handled like telling it it's in a call and talking live, the participants, etc.

Reply:
@crspeller Are you thinking to go as far as letting plugins pass/override a whole prompt template or more like letting them choose which prompt to use and possibly augment existing ones?

Reply:
I am imagining them passing the whole prompt template. That way they can do whatever they want. Ideally they would be able to utilize the default personalities and whatnot as well.